File: core/ui/src/types/spec.ts
import type { SamplingMethod } from "./sampling";

export type SpecType = "MATERIAL" | "IN_PROCESS" | "FINISHED_PRODUCT" | "PACKAGING";
export type SpecStatus = "DRAFT" | "UNDER_REVIEW" | "APPROVED" | "OBSOLETE";
export type TargetMarket =
  | "EU"
  | "US_FDA"
  | "UK_MHRA"
  | "INDIA_CDSCO"
  | "JAPAN_PMDA"
  | "CHINA_NMPA"
  | "WHO_PREQUALIFICATION"
  | "GLOBAL"
  | "INTERNAL";
export type CompendialRef =
  | "PH_EUR"
  | "USP"
  | "BP"
  | "JP"
  | "IP"
  | "CHN_PHARMACOPOEIA"
  | "WHO_INT"
  | "MULTI_COMPENDIAL"
  | "IN_HOUSE"
  | "NONE";
export type ReviewRoute = "QC_ONLY" | "QC_THEN_QA" | "QA_ONLY";

export type SpecParameterTestType =
  | "IDENTITY"
  | "ASSAY"
  | "PURITY"
  | "PHYSICAL"
  | "CHEMICAL"
  | "MICROBIOLOGICAL"
  | "DESCRIPTION";

export type SpecParameterCriteriaType = "NLT" | "NMT" | "RANGE" | "PASS_FAIL" | "COMPLIES" | "TEXT";

export type Spec = {
  id: string;
  specCode: string;
  specName: string;
  revision: string | null;
  specType: SpecType;
  status: SpecStatus;
  samplingMethod: SamplingMethod;
  targetMarket: TargetMarket | null;
  effectiveDate: string | null;
  expiryDate: string | null;
  compendialRef: CompendialRef | null;
  compendialEdition: string | null;
  referenceDocumentNo: string | null;
  referenceAttachment: string | null;
  submittedBy: string | null;
  submittedAt: string | null;
  reviewedBy: string | null;
  reviewedAt: string | null;
  reviewRemarks: string | null;
  reviewRoute: ReviewRoute;
  approvedBy: string | null;
  approvedAt: string | null;
  previousSpecId: string | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type SpecParameter = {
  id: string;
  specId: string;
  parameterName: string;
  testType: SpecParameterTestType;
  moaId: string | null;
  criteriaType: SpecParameterCriteriaType;
  lowerLimit: number | null;
  upperLimit: number | null;
  textCriteria: string | null;
  compendialChapterRef: string | null;
  unit: string | null;
  isMandatory: boolean;
  sequence: number;
  notes: string | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type MaterialSpecLink = {
  id: string;
  materialId: string;
  specId: string;
  isActive: boolean;
  linkedBy: string;
  linkedAt: string;
  delinkedBy: string | null;
  delinkedAt: string | null;
  notes: string | null;
  createdAt: string;
};

export type CreateSpecRequest = {
  specCode: string;
  specName: string;
  revision?: string;
  specType: SpecType;
  samplingMethod: SamplingMethod;
  targetMarket?: TargetMarket;
  effectiveDate?: string;
  expiryDate?: string;
  compendialRef?: CompendialRef;
  compendialEdition?: string;
  referenceDocumentNo?: string;
  referenceAttachment?: string;
  reviewRoute?: ReviewRoute;
  materialIds?: string[];
  createdBy: string;
};

export type ReviewSubmissionRequest = {
  reviewRoute?: ReviewRoute;
};

export type RejectRequest = {
  reviewRemarks: string;
};

export type SpecParameterRequest = {
  parameterName: string;
  testType: SpecParameterTestType;
  moaId?: string;
  criteriaType: SpecParameterCriteriaType;
  lowerLimit?: number;
  upperLimit?: number;
  textCriteria?: string;
  compendialChapterRef?: string;
  unit?: string;
  isMandatory: boolean;
  sequence: number;
  notes?: string;
};

export type LinkMaterialSpecRequest = {
  specId: string;
  notes?: string;
};

export type DelinkMaterialSpecRequest = {
  notes?: string;
};

File: core/src/main/java/com/batchsphere/core/masterdata/spec/entity/Spec.java
package com.batchsphere.core.masterdata.spec.entity;

import com.batchsphere.core.masterdata.quality.enums.CompendialRef;
import com.batchsphere.core.masterdata.quality.enums.ReviewRoute;
import com.batchsphere.core.transactions.sampling.entity.SamplingMethod;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "spec_master")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Spec {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "spec_code", nullable = false, unique = true, length = 100)
    private String specCode;

    @Column(name = "spec_name", nullable = false, length = 255)
    private String specName;

    @Column(name = "revision", length = 50)
    private String revision;

    @Enumerated(EnumType.STRING)
    @Column(name = "spec_type", nullable = false, length = 50)
    private SpecType specType;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private SpecStatus status;

    @Enumerated(EnumType.STRING)
    @Column(name = "sampling_method", nullable = false, length = 50)
    private SamplingMethod samplingMethod;

    @Enumerated(EnumType.STRING)
    @Column(name = "target_market", length = 50)
    private TargetMarket targetMarket;

    @Column(name = "effective_date")
    private LocalDate effectiveDate;

    @Column(name = "expiry_date")
    private LocalDate expiryDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "compendial_ref", length = 50)
    private CompendialRef compendialRef;

    @Column(name = "compendial_edition", length = 100)
    private String compendialEdition;

    @Column(name = "reference_document_no", length = 100)
    private String referenceDocumentNo;

    @Column(name = "reference_attachment", length = 500)
    private String referenceAttachment;

    @Column(name = "submitted_by", length = 100)
    private String submittedBy;

    @Column(name = "submitted_at")
    private LocalDateTime submittedAt;

    @Column(name = "reviewed_by", length = 100)
    private String reviewedBy;

    @Column(name = "reviewed_at")
    private LocalDateTime reviewedAt;

    @Column(name = "review_remarks", columnDefinition = "TEXT")
    private String reviewRemarks;

    @Enumerated(EnumType.STRING)
    @Column(name = "review_route", nullable = false, length = 50)
    private ReviewRoute reviewRoute;

    @Column(name = "approved_by", length = 100)
    private String approvedBy;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "previous_spec_id")
    private UUID previousSpecId;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
